jwst/outlier_detection/utils.py

```diff
@@ -20,177 +26,413 @@
+    Shared code between imaging and spec modes for resampling and median computation
 
     Parameters
     ----------
-    library : ModelLibrary
-        The input data models.
+    input_models : ModelLibrary
+        The input datamodels.
 
-    nsections : int
-        The number of spatial sections in each model
+    resamp : resample.resample.ResampleData object
+        The controlling object for the resampling process.
 
-    section_nrows : int
-        The number of rows in each section
+    maskpt : float
+        The weight threshold for masking out low weight pixels.
 
-    imrows : int
-        The total number of rows in the image
+    save_intermediate_results : bool
+        if True, save the drizzled models and median model to fits.
 
-    Yields
-    ------
-    data_subset : ndarray
-        array of shape (len(library), section_nrows, ncols) representing a spatial
-        subset of all the data arrays in the library
+    make_output_path : function
+        The functools.partial instance to pass to save_median. Must be 
+        specified if save_intermediate_results is True. Default None.
+
+    buffer_size : int
+        The size of chunk in bytes that will be read into memory when computing the median.
+        This parameter has no effect if the input library has its on_disk attribute
+        set to False.
+    """
+    if not resamp.single:
+        raise ValueError("median_with_resampling should only be used for resample_many_to_many")
+    
+    in_memory = not input_models._on_disk
+    indices_by_group = list(input_models.group_indices.values())
+    ngroups = len(indices_by_group)
 
-    weight_subset : ndarray
-        weights corresponding to data_list
+    with input_models:
+        for i, indices in enumerate(indices_by_group):
 
-    row_range : tuple
-        The range of rows in the image covered by the data arrays
+            median_wcs = resamp.output_wcs
+            drizzled_model = resamp.resample_group(input_models, indices)
+
+            if save_intermediate_results:
+                # write the drizzled model to file
+                _fileio.save_drizzled(drizzled_model, make_output_path)
+
+            median_computer = _append_to_median_computer(i, drizzled_model, ngroups, maskpt, in_memory, buffer_size)
+
+    # Perform median combination on set of drizzled mosaics
+    median_data = _evaluate_median_computer(median_computer, in_memory)
+
+    if save_intermediate_results:
+        # Save median model to fits
+        median_model = datamodels.ImageModel(median_data)
+        median_model.update(drizzled_model)
+        median_model.meta.wcs = median_wcs
+        # drizzled model already contains asn_id
+        make_output_path = partial(make_output_path, asn_id=None)
+        _fileio.save_median(median_model, make_output_path)
+    del drizzled_model
+
+    return median_data, median_wcs
+
+
+def _append_to_median_computer(idx, drizzled_model, ngroups, maskpt, in_memory, buffer_size):
```

Comment:
This structure looks more fragile than the previous version - why make median_computer a global here, rather than hold it and append to it in the calling code?

Reply:
Good catch. I missed the use of a global. I agree that having the calling code handle the instance is less fragile.

Reply:
I agree this was not a good idea and I have changed it now

Reply:
That's better, thanks.

I still liked the previous version without the helper functions better, though, for future work supporting median errors in addition to median images.  As written, I could now call separate _make/_add/_eval calls for the error, but it would recompute the weights in the _add call, which seems unnecessary.

Reply:
Without the helper functions, how should I avoid duplication of large portions of that code between the `with_resampling` and `without_resampling` function calls?

Reply:
Thanks, I see your intent now. It looks like maybe 20 lines of code that would be repeated between the two functions, which doesn't seem super concerning to me.  But if you feel strongly about it, can we pull the weight application out of the _add helper?

Reply:
Yeah that sounds good can do

Reply:
Perfect, thank you!